### PR TITLE
OSSM-6383: OSSM in clusterwide mode can include ALL cluster projects

### PR DESCRIPTION
**Pull request by rh-tokeefe**
Version(s): 4.15, 4.14, 4.13, 4.12

Issue:
https://issues.redhat.com/browse/OSSM-6383

Link to docs preview:
https://75873--ocpdocs-pr.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-deployment-models.html#ossm-excluding-namespaces-from-cluster-wide-mesh-console_ossm-deployment-models

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 1]
```diff
@@ -41,12 +46,13 @@ spec:
...
     - matchExpressions:
       - key: kubernetes.io/metadata.name <2>
-        operator: NotIn
+        operator: In
         values:
         - bookinfo
         - httpbin
+        - istio-system
 ----
 <1> Ensures that the mesh discovers namespaces that contain the label `istio-discovery: enabled`. The mesh does not discover namespaces that do not contain the label.
```
Comment (eoinfennessy) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:55]: > The mesh does not discover namespaces that do not contain the label.

This is an inaccuracy. If a namespace matches _any_ of the matches (`matchExpressions` or `matchLabels`), then that namespace will be included in the mesh.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 2]
```diff
@@ -25,7 +25,12 @@ $ oc -n istio-system edit smcp <name> <1>
 ----
 <1> `<name>` represents the name of the `ServiceMeshControlPlane` resource.
 
-. Modify the YAML file so that the `spec.discoverySelectors` field of the `ServiceMeshMemberRoll` resource includes the discovery selector. The following example uses `istio-discovery: enabled`:
+. Modify the YAML file so that the `spec.discoverySelectors` field of the `ServiceMeshMemberRoll` resource includes the discovery selector. In the following example, Istiod discovers any namespace labeled `istio-discovery=enabled` or any namespace with name `bookinfo`, `httpbin` or `istio-system`.
```
Comment (eoinfennessy) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:28]: Maybe we could change this to `istio-discovery: enabled` to remain consistent with the rest of the doc.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 3]
```diff
@@ -25,7 +25,12 @@ $ oc -n istio-system edit smcp <name> <1>
...
++
+[NOTE]
+====
+When you decide which namespaces Istiod can discover, be careful not to include namespaces that may contain sensitive data or services.
```
Comment (eoinfennessy) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:32]: Suggested change: When configuring namespaces that Istiod can discover, be careful not to include namespaces that may contain sensitive services that should not be exposed to the rest of the mesh.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 2 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:28]: done

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 3 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:32]: done

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 1 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:55]: done

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 4]
```diff
@@ -5,8 +5,8 @@
 [id="ossm-excluding-namespaces-from-cluster-wide-mesh-cli_{context}"]
 = Including and excluding namespaces from a cluster-wide mesh by using the CLI
 
-By default, the {SMProductName} Operator uses discovery selectors to identify the namespaces that make up the mesh. Namespaces that do not contain the label defined in the `ServiceMeshMemberRoll` resource are not matched by the discovery selector and are excluded from the mesh.  
- 
+In a cluster-wide mesh, you add discovery selectors to the `ServiceMeshControlPlane` resource. Discovery selectors define which namespaces the control plane can discover. The control plane ignores any namespace that does not match one of the discovery selectors, which excludes the namespace from the mesh.
```
Comment (eoinfennessy) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:8]: > In a cluster-wide mesh, you add discovery selectors...

Adding discovery selectors is optional, but I think this makes it sound like a requirement. I also think that adding something about limiting scope/visibility could give readers a more immediate understanding of what discovery selectors do. How about something like the following?

> In a cluster-wide mesh, discovery selectors can be added to the `ServiceMeshControlPlane` resource. Discovery selectors limit the scope of the control plane to a subset of the namespaces in the cluster. The control plane ignores any namespace that does not match one of the discovery selectors, which excludes the namespace from the mesh.

Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc]: done

**Review by eoinfennessy — APPROVED** [review]
LGTM

**Review by pbajjuri20 — APPROVED** [review]
LGTM..!

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 5]
```diff
@@ -5,8 +5,13 @@
 [id="ossm-excluding-namespaces-from-cluster-wide-mesh-cli_{context}"]
 = Including and excluding namespaces from a cluster-wide mesh by using the CLI
 
-By default, the {SMProductName} Operator uses discovery selectors to identify the namespaces that make up the mesh. Namespaces that do not contain the label defined in the `ServiceMeshMemberRoll` resource are not matched by the discovery selector and are excluded from the mesh.  
- 
+In a cluster-wide mesh, you can add discovery selectors to the `ServiceMeshControlPlane` resource. Discovery selectors define which namespaces the control plane can discover. The control plane ignores any namespace that does not match one of the discovery selectors, which excludes the namespace from the mesh.
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:8]: For sentence 2, is that really a restrictive clause or not? If it is:
```suggestion
In a cluster-wide mesh, you can add discovery selectors to the `ServiceMeshControlPlane` resource. Discovery selectors define the namespaces that the control plane can discover. The control plane ignores any namespace that does not match one of the discovery selectors, which excludes the namespace from the mesh.
```

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 6]
```diff
@@ -5,8 +5,13 @@
...
+
+[NOTE]
+====
+If you install ingress or egress gateways in the control plane namespace, then you must include the control plane namespace in the discovery selectors.
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:12]: ISG prefers that you avoid "If... then" constructions unless you really need them to show a conditional relationship between two long clauses. 

```suggestion
If you install ingress or egress gateways in the control plane namespace, you must include the control plane namespace in the discovery selectors.
```

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 7]
```diff
@@ -25,7 +30,14 @@ $ oc -n istio-system edit smcp <name> <1>
 ----
 <1> `<name>` represents the name of the `ServiceMeshControlPlane` resource.
 
-. Modify the YAML file so that the `spec.discoverySelectors` field of the `ServiceMeshMemberRoll` resource includes the discovery selector. The following example uses `istio-discovery: enabled`:
+. Modify the YAML file so that the `spec.meshConfig` field of the `ServiceMeshControlPlane` resource includes the discovery selector. 
++
+[NOTE]
+====
+When configuring namespaces that `Istiod` can discover, be careful not to include namespaces that might contain sensitive services that should not be exposed to the rest of the mesh.
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:37]:
```suggestion
When configuring namespaces that `Istiod` ${whichever_noun_fits_here} can discover, exclude namespaces that might contain sensitive services that should not be exposed to the rest of the mesh.
```

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 8]
```diff
@@ -25,7 +30,14 @@ $ oc -n istio-system edit smcp <name> <1>
...
+====
++
+In the following example, `Istiod` discovers any namespace labeled `istio-discovery: enabled` or any namespace with the name `bookinfo`, `httpbin` or `istio-system`:
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:40]:
```suggestion
In the following example, `Istiod` ${ISG_wants_a_noun_here_ideally} discovers any namespace that is labeled `istio-discovery: enabled` or any namespace that has the name `bookinfo`, `httpbin` or `istio-system`:
```

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc [hunk 9]
```diff
@@ -41,12 +53,15 @@ spec:
...
         - bookinfo
         - httpbin
+        - istio-system
 ----
-<1> Ensures that the mesh discovers namespaces that contain the label `istio-discovery: enabled`. The mesh does not discover namespaces that do not contain the label.
-<2> Ensures that the mesh does not discover namespaces `bookinfo` and `httpbin`.
+<1> Ensures that the mesh discovers namespaces that contain the label `istio-discovery: enabled`. 
+<2> Ensures that the mesh discovers namespaces `bookinfo`, `httpbin` and `istio-system`.
++
+If a namespace matches any of the `matchExpressions` or `matchLabels`, then the mesh discovers the namespace. The mesh excludes namespaces that do not match either the `matchExpressions` or `matchLabels`.
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:65]:
```suggestion
If a namespace matches any of the `matchExpressions` or `matchLabels` ${plural_noun}, the mesh discovers the namespace. The mesh excludes namespaces that do not match either the `matchExpressions` or `matchLabels`.
```

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc:1]: Same comments here as in the CLI mod where applicable.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc [hunk 10]
```diff
@@ -4,9 +4,14 @@
 :_mod-docs-content-type: PROCEDURE
 [id="ossm-excluding-namespaces-from-cluster-wide-mesh-console_{context}"]
 = Including and excluding namespaces from a cluster-wide mesh by using the web console
- 
-By default, the {SMProductName} Operator uses discovery selectors to identify the namespaces that make up the mesh. Namespaces that do not contain the label defined in the `ServiceMeshMemberRoll` resource are not matched by the discovery selector and are excluded from the mesh.  
- 
+
+In a cluster-wide mesh, you can add discovery selectors to the `ServiceMeshControlPlane` resource. Discovery selectors define which namespaces the control plane can discover. The control plane ignores any namespace that does not match one of the discovery selectors, which excludes the namespace from the mesh.
```
Comment (maxwelldb) [inline modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc:8]: Thought: you probably want to indicate that you're using the web UI here, just as you'd want to indicate that you're using the CLI in the description of the other module.

**Review by maxwelldb — COMMENTED** [review]
Left some comments and suggestions. Nothing major to change. Label for merge review when ready. ⏩

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 5 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:8]: done. made change on console task as well.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 6 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc:12]: done. also made change on console task.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc (hunk 10 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc]: @maxwelldb, the title for each procedures indicates if you are using the console or the CLI. I thought this was sufficient. Can you give me an example of what you had in mind in terms of clarifying this?

Comment (maxwelldb) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc]: I don't typically work on things that have GUIs, but I see this sort of language pretty frequently in the docs for tasks that can be accomplished by using different UIs:

- https://docs.openshift.com/container-platform/4.15/nodes/pods/nodes-pods-autoscaling.html#nodes-pods-autoscaling-creating-web-console_nodes-pods-autoscaling

- https://docs.openshift.com/container-platform/4.15/nodes/pods/nodes-pods-autoscaling.html#nodes-pods-autoscaling-creating-cpu_nodes-pods-autoscaling

Not that pattern verbatim, of course. Just calling out in the description how you're accomplishing the task.

Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-console.adoc]: Thank you, I've made the change. I appreciate your suggestions.

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 7 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc]: done

File: modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc (hunk 8 above)
Comment (rh-tokeefe) [reply modules/ossm-excluding-namespaces-from-cluster-wide-mesh-cli.adoc]: done

**Comment by rh-tokeefe** [comment]
@maxwelldb I've submitted all of the edits and rebased the PR.

**Comment by jeana-redhat** [comment]
> /label merge-review-needed

Hey @rh-tokeefe :) Should I remove the `[WIP]` title and do-not-merge tag here?

**Pull request merged**